File: src/output.rs
use crate::domain::{BudgetStatus, Expense};

pub fn print_add_success(id: i64) {
    println!("Expense added successfully (ID: {id})");
}

pub fn print_expenses(expenses: &[Expense]) {
    if expenses.is_empty() {
        println!("No expenses yet.");
        return;
    }

    let id_w = expenses
        .iter()
        .map(|e| e.id.to_string().len())
        .max()
        .unwrap_or(2)
        .max(2);
    let cat_w = expenses
        .iter()
        .map(|e| e.category.len())
        .max()
        .unwrap_or(8)
        .max(8);
    let desc_w = expenses
        .iter()
        .map(|e| e.description.len())
        .max()
        .unwrap_or(11)
        .max(11);

    println!(
        "{:<id_w$}  {:<10}  {:<cat_w$}  {:<desc_w$}  Amount",
        "ID", "Date", "Category", "Description"
    );
    for e in expenses {
        println!(
            "{:<id_w$}  {:<10}  {:<cat_w$}  {:<desc_w$}  ${:.2}",
            e.id,
            e.date.format("%Y-%m-%d"),
            e.category,
            e.description,
            e.amount
        );
    }
}

/// Prints either a calm "here's where you stand" line or, if the month's
/// spend has gone past its cap, a breach warning -- echoing the language
/// the FIA itself uses when a team overspends the cost cap.
pub fn print_budget_status(status: &BudgetStatus) {
    if status.is_breach() {
        println!(
            "\n🚩 Cost Cap breach for month {:02}: spent ${:.2} against a ${:.2} cap (${:.2} over).",
            status.month,
            status.total,
            status.cap,
            status.total - status.cap
        );
    } else {
        println!(
            "   (month {:02} cost cap: ${:.2} of ${:.2} used, ${:.2} remaining)",
            status.month,
            status.total,
            status.cap,
            status.remaining()
        );
    }
}

File: src/cost_cap_tracker/infrastructure/cli/output.py
"""Presentation-only formatting for expenses and budget status -- mirrors
output.rs, using Rich for semantic-colored terminal output. No business
logic, no I/O beyond the console it's given.
"""

import calendar

from rich.console import Console
from rich.table import Table

from cost_cap_tracker.domain.budget import BudgetStatus
from cost_cap_tracker.domain.expense import Expense

_CLOSE_TO_CAP_RATIO = 0.1


def print_add_success(console: Console, expense_id: int) -> None:
    console.print(f"[green]Expense added successfully (ID: {expense_id})[/green]")


def print_expenses(console: Console, expenses: list[Expense]) -> None:
    if not expenses:
        console.print("[dim]No expenses yet.[/dim]")
        return

    table = Table()
    table.add_column("ID", justify="right")
    table.add_column("Date")
    table.add_column("Category")
    table.add_column("Description")
    table.add_column("Amount", justify="right")
    for expense in expenses:
        table.add_row(
            str(expense.id),
            expense.date.isoformat(),
            expense.category,
            expense.description,
            f"${expense.amount:,.2f}",
        )
    console.print(table)


def print_budget_status(console: Console, status: BudgetStatus) -> None:
    month_name = calendar.month_name[status.month]
    if status.is_breach():
        overage = status.total - status.cap
        console.print(
            f"[bold red]🚩 Cost cap breach for {month_name}: spent "
            f"${status.total:,.2f} of ${status.cap:,.2f} cap "
            f"(over by ${overage:,.2f})[/bold red]"
        )
        return

    remaining = status.remaining()
    style = "yellow" if remaining <= status.cap * _CLOSE_TO_CAP_RATIO else "green"
    console.print(
        f"[{style}]Cost cap for {month_name}: spent ${status.total:,.2f} of "
        f"${status.cap:,.2f} cap (${remaining:,.2f} remaining)[/{style}]"
    )
